<a href="https://colab.research.google.com/github/vjsurampudi/DeepLearningWithTF2/blob/main/NeuralNetworks_with_TF2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [17]:
import tensorflow as tf
print(tf.__version__)

2.4.1


# The Sequential model API

 ## Coding tutorials
 #### [1. Building a Sequential model](#coding_tutorial_1)
 #### [2. Convolutional and pooling layers](#coding_tutorial_2)
 #### [3. The compile method](#coding_tutorial_3)
 #### [4. The fit method](#coding_tutorial_4)
 #### [5. The evaluate and predict methods](#coding_tutorial_5)

***
<a id="coding_tutorial_1"></a>
## Building a Sequential model

In [3]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Flatten, Softmax

#### Build a feedforward neural network model

In [ ]:
# Build the Sequential feedforward neural network model
model = Sequential([
    Flatten(input_shape = (28,28)),       ### This code works even without input shape specified. However weights will not be initialized.
    Dense(16, activation = 'relu'),
    Dense(32, activation = 'relu'),
    Dense(10, activation = 'softmax') 
])

In [ ]:
# Print the model summary
model.weights             

[<tf.Variable 'dense_12/kernel:0' shape=(784, 16) dtype=float32, numpy=
 array([[-0.08234464, -0.04567183,  0.03227976, ..., -0.00182655,
         -0.04649059, -0.08502452],
        [-0.06498432, -0.01981582,  0.03617812, ...,  0.00476246,
          0.04193392,  0.05500585],
        [ 0.08174479,  0.08567253, -0.06953147, ..., -0.06597537,
          0.01573739,  0.03384225],
        ...,
        [ 0.02904293, -0.08103279,  0.08406945, ..., -0.04345918,
          0.07616489,  0.08522111],
        [ 0.04766102,  0.06971447,  0.01367868, ...,  0.08178864,
          0.02156848,  0.03639515],
        [ 0.01446329,  0.01815796, -0.02214409, ...,  0.05119742,
         -0.05197853, -0.06644943]], dtype=float32)>,
 <tf.Variable 'dense_12/bias:0' shape=(16,) dtype=float32, numpy=
 array([0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       dtype=float32)>,
 <tf.Variable 'dense_13/kernel:0' shape=(16, 32) dtype=float32, numpy=
 array([[ 0.0457015 ,  0.25023946,  0.01812416, -0.

Weights will not be initialized until input_shape is provided in model declaration.
Notice the shape is (784, 16) then bias shape (16,0) for first layer 

1.   Shape for first set of weights is (784, 16) followed by bias of (16,0)
2.   Shape for second set of weights is (16, 32) followed by bias of (32,0)
3.   Shape for third set of weights is (32,10) followed by bias of shape (10,0)
4.   Trainable parameters in first dense layer = 784*16 + 16 = 12560

In [ ]:
model.summary() 

Model: "sequential_4"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
flatten_3 (Flatten)          (None, 784)               0         
_________________________________________________________________
dense_12 (Dense)             (None, 16)                12560     
_________________________________________________________________
dense_13 (Dense)             (None, 32)                544       
_________________________________________________________________
dense_14 (Dense)             (None, 10)                330       
Total params: 13,434
Trainable params: 13,434
Non-trainable params: 0
_________________________________________________________________


In [ ]:
### Naming the layers and substituting Softmax activation function in the last dense layer by a new dense layer
model = Sequential([
    Flatten(input_shape = (28,28)),      
    Dense(16, activation = 'relu', name= 'layer_1'),
    Dense(32, activation = 'relu'),
    Dense(10),
    Softmax() 
])

In [ ]:
model.summary()

Model: "sequential_5"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
flatten_5 (Flatten)          (None, 784)               0         
_________________________________________________________________
layer_1 (Dense)              (None, 16)                12560     
_________________________________________________________________
dense_17 (Dense)             (None, 32)                544       
_________________________________________________________________
dense_18 (Dense)             (None, 10)                330       
_________________________________________________________________
softmax (Softmax)            (None, 10)                0         
Total params: 13,434
Trainable params: 13,434
Non-trainable params: 0
_________________________________________________________________


***
<a id="coding_tutorial_2"></a>
## Convolutional and pooling layers

In [5]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Flatten, Conv2D, MaxPooling2D

#### Build a convolutional neural network model

In [ ]:
# Build the Sequential convolutional neural network model
model = Sequential([
                    Conv2D(16, (3,3), activation='relu', input_shape = (28,28,1)),
                    MaxPooling2D((3,3)),
                    Flatten(),
                    Dense(10, activation = 'softmax')
])

1. Number of channels in input shape (1 in the above example) is mandatory
2. None is the default first parameter in 'Output Shape' below. It indicates batch size.

In [ ]:
# Print the model summary

model.summary()

Model: "sequential_1"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_2 (Conv2D)            (None, 26, 26, 16)        160       
_________________________________________________________________
max_pooling2d_1 (MaxPooling2 (None, 8, 8, 16)          0         
_________________________________________________________________
flatten_1 (Flatten)          (None, 1024)              0         
_________________________________________________________________
dense_1 (Dense)              (None, 10)                10250     
Total params: 10,410
Trainable params: 10,410
Non-trainable params: 0
_________________________________________________________________


Note that the output shape of conv2d_2 layer is different (and did not change from input shape) when the padding is set as 'SAME'

Use the parameter data_format = 'channels_first' if the you want to input channels first in the input shape

In [ ]:
# Build the Sequential convolutional neural network model
model = Sequential([
                    Conv2D(16, (3,3), activation='relu', padding = 'SAME', input_shape = (1,28,28), data_format = 'channels_first'),
                    MaxPooling2D((3,3)),
                    Flatten(),
                    Dense(10, activation = 'softmax')
])

In [ ]:
model.summary()

Model: "sequential_6"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_7 (Conv2D)            (None, 16, 28, 28)        160       
_________________________________________________________________
max_pooling2d_6 (MaxPooling2 (None, 5, 9, 28)          0         
_________________________________________________________________
flatten_6 (Flatten)          (None, 1260)              0         
_________________________________________________________________
dense_6 (Dense)              (None, 10)                12610     
Total params: 12,770
Trainable params: 12,770
Non-trainable params: 0
_________________________________________________________________


Note the changes in output shape of Conv2d layer when stride is changed to 2. 

In [6]:
# Build the Sequential convolutional neural network model
model = Sequential([
                    Conv2D(16, (3,3), activation='relu', padding = 'SAME', strides = 2, input_shape = (28,28,1)),
                    MaxPooling2D((3,3)),
                    Flatten(),
                    Dense(10, activation = 'softmax')
])

In [7]:
model.summary()

Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d (Conv2D)              (None, 14, 14, 16)        160       
_________________________________________________________________
max_pooling2d (MaxPooling2D) (None, 4, 4, 16)          0         
_________________________________________________________________
flatten (Flatten)            (None, 256)               0         
_________________________________________________________________
dense (Dense)                (None, 10)                2570      
Total params: 2,730
Trainable params: 2,730
Non-trainable params: 0
_________________________________________________________________


***
<a id="coding_tutorial_3"></a>
## The compile method

#### Compile the model

In [23]:
# Define the model optimizer, loss function and metrics
model.compile(
    optimizer='adam', #sgd, rmsprop
    loss = 'sparse_categorical_crossentropy', #categorical_crossentropy
    metrics = ['accuracy','mae'] #mae
)


In [24]:
# Print the resulting model attributes
print(model.loss)
print(model.optimizer)
print(model.metrics)

sparse_categorical_crossentropy
[]


In [20]:
opt = tf.keras.optimizers.Adam(learning_rate = 0.005)
acc = tf.keras.metrics.SparseCategoricalAccuracy()
mae = tf.keras.metrics.MeanAbsoluteError()

model.compile(
    optimizer=opt,
    loss = 'categorical_crossentropy',
    metrics = [acc,'mae'] 
)

In [22]:
# Print the resulting model attributes
print(model.loss)
print(model.optimizer)
print(model.metrics)
print(model.optimizer.lr)

categorical_crossentropy
[]
<tf.Variable 'learning_rate:0' shape=() dtype=float32, numpy=0.005>


'***
<a id="coding_tutorial_4"></a>
### The fit method

In [ ]:
from tensorflow.keras.preprocessing import image
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd

#### Load the data

In [ ]:
# Load the Fashion-MNIST dataset

fashion_mnist_data = tf.keras.datasets.fashion_mnist
(train_images, train_labels), (test_images, test_labels) = fashion_mnist_data.load_data()

In [ ]:
# Print the shape of the training data



In [ ]:
# Define the labels

labels = [
    'T-shirt/top',
    'Trouser',
    'Pullover',
    'Dress',
    'Coat',
    'Sandal',
    'Shirt',
    'Sneaker',
    'Bag',
    'Ankle boot'
]

In [ ]:
# Rescale the image values so that they lie in between 0 and 1.



In [ ]:
# Display one of the images



#### Fit the model

In [ ]:
# Fit the model



#### Plot training history

In [ ]:
# Load the history into a pandas Dataframe



In [ ]:
# Make a plot for the loss



In [ ]:
# Make a plot for the accuracy



In [ ]:
# Make a plot for the additional metric



***
<a id="coding_tutorial_5"></a>
## The evaluate and predict methods

In [ ]:
import matplotlib.pyplot as plt
import numpy as np

#### Evaluate the model on the test set

In [ ]:
# Evaluate the model



#### Make predictions from the model

In [ ]:
# Choose a random test image

random_inx = np.random.choice(test_images.shape[0])

test_image = test_images[random_inx]
plt.imshow(test_image)
plt.show()
print(f"Label: {labels[test_labels[random_inx]]}")

In [ ]:
# Get the model predictions

